### ISSUE TITLE
Replace earlier text in page_script

### ISSUE THREAD

**Issue by rudiedirkx** [dompdf/dompdf#1506]
In the header of the first page, I want to show the number of total pages. I can do that with

```
<script type="text/php">
	$pdf->page_script('
   		if ($PAGE_NUM == 1) {
   			$font = $fontMetrics->getFont("sans-serif");
   			$size = 8;
   			$y = 216;
   			$x = $pdf->get_width() - 138;
   			$pdf->text($x, $y, $PAGE_COUNT, $font, $size);
   		}
	');
</script>
```

but `216` and `138` are magic numbers, and the font size and family and position etc can all change within the document, so I don't like this method.

Is there a way to print a marker and then replace it with a page_script? Or maybe with objects? Or something else that's readable?

**Comment by bsweeney** [comment]
It is possible, but not simple. Have a look at this TOC sample: http://eclecticgeek.com/dompdf/debug.php?identifier=toc-v0.6.x

**Comment by rudiedirkx** [comment]
Yes, that is somewhat clunky, but with a little bit of love, that could be perfect. Maybe with support for pre- and post-processing.

This does exactly what I want though. I just have to be careful with the variable names and content. If one is much longer than the other, it's rendered completely wrong, but that's to be expected. Works perfectly for short replacements (and probably fixed/absolute positions).

If I think of something useful pre-/post-processor-wise, I'll create a PR. Or is that a thing already?

**Comment by rudiedirkx** [comment]
Actually it's less perfect than previous reports... =)

For an object only on page 1, it works perfectly, because the placeholders (`%N%` and `%M%` in my case) are only replaced once.

For objects on every other page, replacement only works once, so every next page will have `page 2 of 5`, because this replacement happened on page 2 first... Changing the object more than once, changes it for every page, because it's the same object. I can't seem to reuse the object... Any ideas?

**Comment by bsweeney** [comment]
Are you using fixed-position elements for that other content? If so I can see why that would be a problem. Can you post a sample of your document?

**Comment by bsweeney** [comment]
I see you came back in #1512. Presumably I can test with that document?

FYI, since the current page can be generated using CSS you should be able to change up your header like so:

```
	<header class="page-all">
		17-18/000105
		&nbsp;
		<script xtype="text/php">$GLOBALS['pager_n'] = $pdf->open_object();</script>
		pagina <span class="pagenum"></span> van %m%
		<script xtype="text/php">$pdf->close_object();</script>
	</header>
```

with this additional CSS:

```
.pagenum:before { content: counter(page); }
```

Then your script at the end only needs to deal with the total pages replacement. I'm not sure when we'll address being able to use CSS to generate the page count, but it is on our radar.

**Comment by rudiedirkx** [comment]
Yes, the #1512 document is the one.

I'll try the `counter(page)` method, but I still like the replacement, also inside a fixed element. Is there a way to clone these objects? Can I just copy the array to a new object number?

**Comment by rudiedirkx** [comment]
This works for multiple pages:

	<script type="text/php">
		$pdf->page_script('
			if ($PAGE_NUM == 1 ) {
				$pdf->get_cpdf()->objects[ $GLOBALS["pager_1"] ]["c"] = strtr($pdf->get_cpdf()->objects[ $GLOBALS["pager_1"] ]["c"], [
					"%n%" => $PAGE_NUM,
					"%m%" => $PAGE_COUNT,
				]);
				$pdf->add_object($GLOBALS["pager_1"]);
			}
			else {
				$nobj = $pdf->open_object(); $pdf->close_object();
				$pdf->get_cpdf()->objects[$nobj] = $pdf->get_cpdf()->objects[ $GLOBALS["pager_n"] ];
				$pdf->get_cpdf()->objects[$nobj]["c"] = strtr($pdf->get_cpdf()->objects[$nobj]["c"], [
					"%n%" => $PAGE_NUM,
					"%m%" => $PAGE_COUNT,
				]);
				$pdf->add_object($nobj);
			}
		');
	</script>

but it's not very elegant. Could be more elegant if objects were more OO.

Maybe create a `$pdf->clone_object()` method? Makes cleaner code...

**Comment by bsweeney** [comment]
Nice hacking there! I'm glad you were able to work it out.

The CPDF class is in need of a major rewrite, but it's not on our immediate to-do list.

**Comment by TechGameTeddy** [comment]
Finding effective use of Wordpress, Vagrant, Puppet, and TurnkeyLinux

**Comment by rudiedirkx** [comment]
Is there a rewrite/refactor issue to follow? It's still up to you, but I'd like to follow progress =)

**Comment by bsweeney** [comment]
Not really. The most relevant may be #184. As you can see the idea has been around since before we moved to git.

**Comment by shaiby** [comment]
Can some one help me whats I am doing wrong while generating TOC. Page number are not replacing.
http://eclecticgeek.com/dompdf/debug.php?identifier=cd82468c0a99e995139bf3235162bfc4
I think it will require use of CPDF but I am not sure how to use it

**Comment by bsweeney** [comment]
@shaiby there's an example here: https://github.com/dompdf/dompdf/issues/1756#issuecomment-1006718311
